audio/out/ao_audiotrack.c

```diff
@@ -564,21 +565,33 @@ static MP_THREAD_VOID ao_thread(void *arg)
             int64_t ts = mp_time_ns();
             ts += MP_TIME_S_TO_NS(read_samples / (double)(ao->samplerate));
             ts += MP_TIME_S_TO_NS(AudioTrack_getLatency(ao));
-            int samples = ao_read_data_nonblocking(ao, &p->chunk, read_samples, ts);
-            int ret = AudioTrack_write(ao, samples * ao->sstride);
-            if (ret >= 0) {
-                p->written_frames += ret / ao->sstride;
-            } else if (ret == AudioManager.ERROR_DEAD_OBJECT) {
-                MP_WARN(ao, "AudioTrack.write failed with ERROR_DEAD_OBJECT. Recreating AudioTrack...\n");
-                if (AudioTrack_Recreate(ao) < 0) {
-                    MP_ERR(ao, "AudioTrack_Recreate failed\n");
+            bool eof;
+            int samples = ao_read_data(ao, &p->chunk, read_samples, ts, &eof, false, false);
+            if (samples) {
+                after_eof = false;
+                int ret = AudioTrack_write(ao, samples * ao->sstride);
+                if (ret >= 0) {
+                    p->written_frames += ret / ao->sstride;
+                } else if (ret == AudioManager.ERROR_DEAD_OBJECT) {
+                    MP_WARN(ao, "AudioTrack.write failed with ERROR_DEAD_OBJECT. Recreating AudioTrack...\n");
+                    if (AudioTrack_Recreate(ao) < 0) {
+                        MP_ERR(ao, "AudioTrack_Recreate failed\n");
+                    }
+                } else {
+                    MP_ERR(ao, "AudioTrack.write failed with %d\n", ret);
                 }
             } else {
-                MP_ERR(ao, "AudioTrack.write failed with %d\n", ret);
+                if (eof) {
+                    after_eof = true;
+                    ao_stop_streaming(ao);
+                }
+                if (!after_eof) {
+                    continue;
+                }
             }
-        } else {
-            mp_cond_timedwait(&p->wakeup, &p->lock, MP_TIME_MS_TO_NS(300));
         }
+
+        mp_cond_timedwait(&p->wakeup, &p->lock, MP_TIME_MS_TO_NS(300));
```

Comment:
wouldn't that mean it sleeps after every write?

Reply:
If `after_eof` is `false`, the loop will just continue. See lines above:

```c
                if (!after_eof) {
                    continue;
                }
```